File: heap/Solution264.java
import java.util.HashSet;
import java.util.Set;
import java.util.Queue;
import java.util.PriorityQueue;
import java.util.Comparator;

class Solution264 {
   public int nthUglyNumber(int n) {
        Set<Long> set = new HashSet<>();

        Queue<Long> queue = new PriorityQueue<>();
        queue.add((long) 1);
        set.add((long) 1);
        int cur = 0;
        while(cur < n){
            long number = queue.poll();
            cur += 1;
            if(cur == n){
                return (int) number;
            }
            long n1 = number * 2;
            long n2 = number * 3;
            long n3 = number * 5;
            if(!set.contains(n1)){
                queue.add(n1);
            }
            if(!set.contains(n2)){
                queue.add(n2);
            }
            if(!set.contains(n3)){
                queue.add(n3);
            }
            set.add(n1);
            set.add(n2);
            set.add(n3);
        }
        return -1;
    }

    public static void main(String[] args){
    	Solution264 s = new Solution264();
    	System.out.println(s.nthUglyNumber(2));
    }
}

File: C++/Solution264.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<stack>
#include<queue>

using ll = long long;
using namespace std;

class Solution264 {
public:
    int nthUglyNumber(int n) {
        unordered_set<ll> st;
        vector<ll> vec(1691);
        priority_queue<ll, vector<ll>, greater<ll>> pq;
        int times[] = {2, 3, 5};
        pq.push(1);
        st.insert(1);
        for(int i = 1; i <= 1690; i++){
            ll cur = pq.top();
            //cout << cur << endl;
            pq.pop();
            vec[i] = cur;
            for(int j = 0; j < 3; j++){
                ll next = cur * times[j];
                if(st.count(next)){
                    continue;
                }
                st.insert(next);
                pq.push(next);
            }
        }
        return vec[n];
    }
};

int main(){
    priority_queue<int, vector<int>, greater<int>> pq;
    pq.push(1);
    pq.push(2);
    while(!pq.empty()){
        cout << pq.top() << endl;
        pq.pop();
    }
    return 0;
}
